R/boxr_auth.R

```diff
@@ -386,3 +386,169 @@ box_auth_on_attach <- function(auth_on_attach = FALSE) {
+#' @importFrom jose jwt_claim jwt_encode_sig
+#' @importFrom rlang %||%
+#' @importFrom glue glue
+#' @export
+box_auth_jwt <- function(user_id = NULL, config_file = NULL,
+                         write.Renv) {
+  
+  # deprecate write.Renv
+  if (!missing(write.Renv)) {
+    warning(
+      "argument `write.Renv` is deprecated; ",
+      "you must edit your Renviron file directly."
+    )
+  }
+  
+  user_id_env <- Sys.getenv("BOX_USER_ID")
+  config_file_env <- Sys.getenv("BOX_CONFIG_FILE")
+  
+  # if no input, look to .Renviron for the id and secret
+  if (is_void(user_id) && !is_void(user_id_env)) {
+    message("Using `BOX_USER_ID` from environment")
+    user_id <- user_id_env
+  }
+  
+  if (is_void(config_file) && !is_void(config_file_env)) {
+    message("Using `BOX_CONFIG_FILE` from environment")
+    config_file <- config_file_env 
+  }
+  if (is_void(user_id) || is_void(config_file)) {
+    stop(
+      "box.com authorization no possible; user_id and/or config_file not found\n",
+      "See ?box_auth_jwt for help."
+    )
+  }
+  config <- jsonlite::fromJSON(config_file)
+  # de-crypt the key
+  key <- openssl::read_key(config$boxAppSettings$appAuth$privateKey,
+                           config$boxAppSettings$appAuth$passphrase)
+  # build out a claim/payload as a specific user
+  auth_url <- "https://api.box.com/oauth2/token"
+  claim <- jose::jwt_claim(
+    iss = config$boxAppSettings$clientID,
+    sub = as.character(user_id), # maybe don't need this?
+    box_sub_type = "user",
+    aud = auth_url,
+    jti = openssl::base64_encode(openssl::rand_bytes(16)),
+    exp = unclass(Sys.time()) + 45
+  )
+  # sign claim with key
+  assertion <- jose::jwt_encode_sig(
+    claim, key,
+    header = list("kid" = config$boxAppSettings$appAuth$publicKeyID)
+    )
+  params <- list(
+    "grant_type" = "urn:ietf:params:oauth:grant-type:jwt-bearer",
+    "assertion" = assertion,
+    "client_id"     = config$boxAppSettings$clientID,
+    "client_secret"  = config$boxAppSettings$clientSecret
+  )
+  req <- httr::POST(auth_url,
+                    body = params,
+                    encode = "form")
+  box_token <- httr::content(req)$access_token
+  
+  # Test the connection; retrieve the username
+  box_token_bearer <- httr::add_headers(Authorization = paste("Bearer", box_token))
+  test_req <- httr::GET("https://api.box.com/2.0/folders/0", box_token_bearer)
+  if (httr::http_status(test_req)$cat != "Success") {
+    stop("Login at box.com failed; unable to connect to API.")
+  }
+  cr <- httr::content(test_req)
+  
+  # Write to options
+  options(
+    boxr.token = NULL, # wipe any token set by box_auth() to prevent auth confusion in POST operations
+    boxr_token_jwt = box_token_bearer,
+    boxr.username = cr$owned_by$login,
+    box_wd = "0"
+  )
+  
+  message(
+    paste0(
+      "boxr: Authenticated at box.com as ",
+      cr$owned_by$name, " (",
+      cr$owned_by$login, ")"
+    )
+  )
+  
+  # new_jwt_info <- 
```

Comment:
I imagine that this is commented out while we sort out what we want to do to help the user add things to their `.Renviron` file. 

In terms of what we want to do, how will this be different from the OAuth2 case?

Reply:
I think it is identical, not sure why it's commented...